File: langs/node/test/conformance.test.js
/**
 * Suite de conformidade do port Node — roda os vetores de
 * `conformance/vectors.json` e compara byte-a-byte com `expected_md5` da
 * referência Python (`conformance/reference.py`).
 *
 * Usa `node:test` (runner nativo) para não exigir mocha/jest/vitest.
 * Resolução de caminhos via `import.meta.url` (independe do CWD do runner).
 */

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFile, readFileSync } from 'node:fs';
import { readFile as readFileAsync } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import { dirname, join, resolve } from 'node:path';

import {
  hashTiss,
  hashTissFile,
  InvalidTissXmlError,
  TISS_NAMESPACE,
} from '../src/index.js';

const here = dirname(fileURLToPath(import.meta.url));
// langs/node/test → langs/node → langs → raiz → conformance
const conformanceDir = resolve(here, '..', '..', '..', 'conformance');
const vectorsPath = join(conformanceDir, 'vectors.json');

// Carga síncrona para definir testes dinamicamente (uma chamada `test()`
// por vetor, com nome bom no relatório).
const manifestRaw = readFileSync(vectorsPath, 'utf8');
const manifest = JSON.parse(manifestRaw);

for (const vec of manifest.vectors) {
  test(`conformance ${vec.id} — ${vec.desc}`, async () => {
    const inputPath = join(conformanceDir, vec.input);
    const bytes = await readFileAsync(inputPath);
    const got = hashTiss(bytes);
    assert.equal(
      got,
      vec.expected_md5,
      `hash divergente para ${vec.id}: got=${got} expected=${vec.expected_md5}`,
    );
  });
}

test('TISS_NAMESPACE exporta a URI correta', () => {
  assert.equal(
    TISS_NAMESPACE,
    'http://www.ans.gov.br/padroes/tiss/schemas',
  );
});

test('hashTissFile bate com hashTiss(read)', async () => {
  const inputPath = join(conformanceDir, 'inputs/syn_acento.xml');
  const fromFile = await hashTissFile(inputPath);
  const fromBytes = hashTiss(await readFileAsync(inputPath));
  assert.equal(fromFile, fromBytes);
  // Hash conhecido da referência: prova UTF-8 nos bytes do MD5.
  assert.equal(fromFile, 'a20afc9a89aadaa2179d03d225337662');
});

test('hashTiss lança TypeError se input não for bytes', () => {
  assert.throws(
    () => hashTiss('string ao invés de bytes'),
    TypeError,
  );
  assert.throws(
    () => hashTiss({ foo: 'bar' }),
    TypeError,
  );
});

test('hashTiss lança InvalidTissXmlError em XML mal-formado', () => {
  const malformed = Buffer.from('<root><sem-fechar>');
  assert.throws(
    () => hashTiss(malformed),
    (err) => err instanceof InvalidTissXmlError,
  );
});

test('hashTiss aceita Uint8Array (não só Buffer)', () => {
  // Pega um XML válido pequeno e converte pra Uint8Array puro.
  const xml = `<?xml version='1.0' encoding='utf-8'?>` +
    `<ans:mensagemTISS xmlns:ans="${TISS_NAMESPACE}">` +
    `<ans:epilogo><ans:hash>LIXO</ans:hash></ans:epilogo>` +
    `</ans:mensagemTISS>`;
  const buf = Buffer.from(xml, 'utf8');
  const u8 = new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);
  // Sem conteúdo de folha além do <ans:hash> zerado → MD5("") esperado.
  assert.equal(
    hashTiss(u8),
    'd41d8cd98f00b204e9800998ecf8427e',
  );
});

// Sanidade adicional: comentário XML ENTRA no concat (ambiguidade #2 da
// referência). Já coberta indiretamente pelo vetor syn_comentario.xml, mas
// vale teste explícito que documenta a intenção.
test('comentário XML contribui textContent ao concat (ambiguidade #2)', () => {
  // Documento minimalista com um único comentário antes do <ans:hash>.
  // Sem o comentário, o hash seria MD5("") = d41d8cd9...
  // Com o comentário, o hash muda — basta provar que ≠ MD5("").
  const xml = `<?xml version='1.0' encoding='utf-8'?>` +
    `<ans:mensagemTISS xmlns:ans="${TISS_NAMESPACE}">` +
    `<ans:cabecalho><!-- COMENTARIO --></ans:cabecalho>` +
    `<ans:epilogo><ans:hash></ans:hash></ans:epilogo>` +
    `</ans:mensagemTISS>`;
  const h = hashTiss(Buffer.from(xml));
  assert.notEqual(h, 'd41d8cd98f00b204e9800998ecf8427e');
});

// Suppress unused warning para `readFile` callback variant.
void readFile;

File: langs/dart/test/conformance_test.dart
// Suite de conformidade do port Dart: roda os vetores de
// `conformance/vectors.json` e compara byte-a-byte com `expected_md5` da
// referencia Python (`conformance/reference.py`).
//
// Resolucao de caminhos via Platform.script (independe do CWD do runner).

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:tiss_hash/tiss_hash.dart';

/// Resolve o diretorio `conformance/` subindo a partir do CWD (normalmente a
/// raiz do package `langs/dart` quando se roda `dart test`) ate achar um dir
/// `conformance/` irmao. Independe de onde o runner foi invocado.
Directory _conformanceDir() {
  var dir = Directory.current;
  for (var i = 0; i < 8; i++) {
    final candidate = Directory('${dir.path}/conformance');
    if (candidate.existsSync() &&
        File('${candidate.path}/vectors.json').existsSync()) {
      return candidate;
    }
    final parent = dir.parent;
    if (parent.path == dir.path) break; // chegou na raiz do FS
    dir = parent;
  }
  throw StateError(
    'diretorio conformance/ nao encontrado subindo a partir de '
    '${Directory.current.path}; rode os testes dentro do checkout completo '
    'do repo lib_hash_ans.',
  );
}

void main() {
  final conformanceDir = _conformanceDir();
  final vectorsFile = File('${conformanceDir.path}/vectors.json');

  test('vectors.json existe no conformance/', () {
    expect(vectorsFile.existsSync(), isTrue,
        reason: 'vectors.json ausente: ${vectorsFile.path}');
  });

  final manifest =
      jsonDecode(vectorsFile.readAsStringSync()) as Map<String, dynamic>;
  final vectors = (manifest['vectors'] as List).cast<Map<String, dynamic>>();

  group('conformidade (vectors.json)', () {
    for (final vec in vectors) {
      final id = vec['id'] as String;
      final desc = vec['desc'] as String? ?? '';
      // Campo `expect`: ausente ou "hash" = POSITIVO (compara expected_md5);
      // "error" = NEGATIVO (port DEVE rejeitar, nao produzir hash).
      final isNegative = vec['expect'] == 'error';

      test('$id — $desc', () {
        final inputPath = '${conformanceDir.path}/${vec['input']}';
        final bytes = File(inputPath).readAsBytesSync();

        if (isNegative) {
          expect(
            () => hashTiss(bytes),
            throwsA(isA<InvalidTissXmlException>()),
            reason: 'vetor negativo $id deveria lancar InvalidTissXmlException',
          );
          return;
        }

        final got = hashTiss(bytes);
        expect(
          got,
          equals(vec['expected_md5']),
          reason: 'hash divergente para $id',
        );
      });
    }
  });

  group('API auxiliar', () {
    test('tissNamespace exporta a URI correta', () {
      expect(tissNamespace, 'http://www.ans.gov.br/padroes/tiss/schemas');
    });

    test('hashTissFile bate com hashTiss(bytes)', () async {
      final inputPath = '${conformanceDir.path}/inputs/syn_acento.xml';
      final fromFile = await hashTissFile(inputPath);
      final fromBytes = hashTiss(File(inputPath).readAsBytesSync());
      expect(fromFile, fromBytes);
      // Hash conhecido da referencia: prova UTF-8 nos bytes do MD5.
      expect(fromFile, 'a20afc9a89aadaa2179d03d225337662');
    });

    test('XML mal-formado lanca InvalidTissXmlException', () {
      final malformed = utf8.encode('<root><sem-fechar>');
      expect(
        () => hashTiss(malformed),
        throwsA(isA<InvalidTissXmlException>()),
      );
    });

    test('documento sem <ans:hash> e valido (concatena tudo)', () {
      final xml = utf8.encode(
        "<?xml version='1.0' encoding='utf-8'?>"
        '<ans:mensagemTISS xmlns:ans="$tissNamespace">'
        '<ans:cabecalho><ans:id>X</ans:id></ans:cabecalho>'
        '</ans:mensagemTISS>',
      );
      // Sem hash, so a folha <ans:id>X</ans:id> contribui -> MD5("X").
      expect(hashTiss(xml), '02129bb861061d1a052c592e2dc6b383');
    });

    test('namespace TISS como default (sem prefixo) localiza <hash>', () {
      // hash com namespace default deve ser zerado -> so folhas restantes contam.
      final xml = utf8.encode(
        "<?xml version='1.0' encoding='utf-8'?>"
        '<mensagemTISS xmlns="$tissNamespace">'
        '<epilogo><hash>LIXO</hash></epilogo>'
        '</mensagemTISS>',
      );
      // <hash> zerado, nenhuma outra folha -> MD5("").
      expect(hashTiss(xml), 'd41d8cd98f00b204e9800998ecf8427e');
    });

    test('comentario XML contribui ao concat (ambiguidade #2)', () {
      // Sem o comentario, hash seria MD5("") = d41d8cd9...
      final xml = utf8.encode(
        "<?xml version='1.0' encoding='utf-8'?>"
        '<ans:mensagemTISS xmlns:ans="$tissNamespace">'
        '<ans:cabecalho><!-- COMENTARIO --></ans:cabecalho>'
        '<ans:epilogo><ans:hash></ans:hash></ans:epilogo>'
        '</ans:mensagemTISS>',
      );
      expect(hashTiss(xml), isNot('d41d8cd98f00b204e9800998ecf8427e'));
    });
  });
}
